modules/Bio/EnsEMBL/BaseAlignFeature.pm

```diff
@@ -1009,7 +1012,303 @@ sub _query_unit {
   throw( "Abstract method call!" );
 }
 
+=head2 _mdtag_alignment_length
+
+  Arg [1]    : None
+  Description: return the alignment length (including indels) based on the mdtag (mdz) string
+  Returntype : int
+  Exceptions : none
+  Caller     : internal
+  Status     : Stable
+
+=cut
+
+sub _mdtag_alignment_length {
+  my $self = shift;
+
+  if (! defined $self->{'_alignment_length'} && defined $self->cigar_string) {
+    my $mdz_string =  $self->cigar_string;
+    my $chunks = $self->_get_mdz_chunks($mdz_string);
+    my $alignment_length = 0;
+    $alignment_length = $self->_get_mdz_alignment_length($chunks);
+    $self->{'_alignment_length'} = $alignment_length;
+  }
+  return $self->{'_alignment_length'};
+}
+
+=head2 _get_mdz_chunks
+
+  Arg [1]    : mdz string
+  Description: parses the mdz string and group it according the type
+  Returntype : array of strings
+  Exceptions : none
+  Caller     : internal
+  Status     : Stable
+
+=cut
+
+sub _get_mdz_chunks {
+  my $self = shift;
+
+  my $mdz_string = shift;
+  my $alignment_length = 0;
+  my @chunks;
+    if ( $mdz_string =~ /^MD:Z:(.*)/ ){
+      my $mdtag = $1;
+
+      #if start to end is a number then all match return as it is
+      if($mdtag =~/^\d+$/){
+        return [$mdtag];
+      }else{
+        my @char_arr = split('',$mdtag);
+        for(my $i=0; $i< scalar(@char_arr); $i++){
+        my $char = $char_arr[$i];
+
+        #determine the type and next type
+        my $type = $self->_get_mdz_chunk_type($char);
+        my $next_char = $char_arr[++$i];
+        my $next_type = $self->_get_mdz_chunk_type( $next_char);
+
+        my $chunk = $char;
+
+        #group chars in mdz based on the type, break if it differenc and push to chunks
+        while($type eq $next_type and $i <= $#char_arr){
```

Comment:
I think you might be over thinking this with a lot of ++ and -- index variable. So basically you want to start a new bucket every time the "type" changes (^, digit, alpha). You haven an array with all the character split up. You could go through popping off that array, .= the next character on to the current "chunk" bucket unless the class changes. Otherwise, push the current bucket on to the pile and start another. It would likely be a little less cluttered.